README.md
# ChatGPT Client

Access OpenAI ChatGPT Official API using Dart Language. Supports any Dart project and all Flutter target platforms (iOS, Android, Windows, Linux, Web)

## Features

- Using OpenAI ChatGPT Official Completions API Endpoint.
- Maintain Chat History List on the client (~4000 tokens) so new prompt is aware of the previous chat context.
- Support any custom model as Parameter (GPT-3.5, GPT-4, etc)
- Support Stream HTTP Response using native Dart Stream.
- Support Standard HTTP Response.
- Pass custom system prompt and temperature.

## Getting started

Register for API key from [OpenAI](https://openai.com/api). 

## Installation

This package is hosted in [Pub Dev Library](https://pub.dev/packages/chatgpt_client)

### Flutter App

```dart
flutter pub add chagpt_client
```

### Dart App
```dart
dart pub add chagpt_client
```

## Usage

### Initialization

Initialize with api key. Default model is `gpt-3.5-turbo`.

```dart
import 'package:chatgpt_client/chatgpt_client.dart';

const client = ChatGPTClient(apiKey: "API_KEY");
```

optionally, you can provide the system prompt, temperature, and model like so.

```dart
const client = ChatGPTClient(apiKey: "API_KEY",
    model: "gpt-4",
    systemPrompt: "You are a CS Professor",
    temperature: 0.7);
```

There are 2 APIs: stream and normal

### Stream Response

The server will yield chunks of data until the stream completes or throws an error.

```dart
try {
    var text = "";
    final stream = client.sendMessageStream(prompt);
    await for (final textChunk in stream) {
        text += textChunk;
        print(textChunk);
    }
    print(text);
} catch (exception) {
    print(exception.toString());
}
```

### Normal
A normal HTTP request and response lifecycle. Server will send the complete text (it will take more time to response)

```dart
try {
    final text = await client.sendMessage(prompt);
    print(text);
} catch (exception) {
    print(exception.toString());
}   
```

## History List

The client stores the history list of the conversation that will be included in the new prompt so ChatGPT aware of the previous context of conversation. When sending new prompt, the client will make sure the token is not exceeding 4000 (using calculation of 1 token=4chars), in case it exceeded the token, some of previous conversations will be truncated

You can also delete the history list by invoking

```dart
client.clearHistoryList();
```

CHANGELOG.md
## 1.1.1

* Fix Stream issue in Safari browser, for now it will wait for chunks to be completed. This will be fixed when `dart:http` fix the streamed response issue in Web

## 1.1.0

* Fix HTTP Stream on Web

## 1.0.0

* Bump version to stable `1.0.0` release

## 0.0.7

* Update Pubspec and readme

## 0.0.6

* Handle parsing case where stream response has multiple chunks of `data:` in byte stream

## 0.0.5

* Use substring instead of replaceFirst for parsing stream response

## 0.0.4

* Fix parsing issue in stream response


## 0.0.3

* Rename `ChatGPTAPI` class to `ChatGPTClient` for consistency.


## 0.0.2

* Update README


## 0.0.1

* Initial release of ChatGPT API Client


lib/src/chatgptclient.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:chatgpt_client/src/message.dart';
import 'stream_client.dart';

/// A class to interact with OpenAI ChatGPT Completions API
/// Support various models such as gpt-3.5-turbo, gpt-4, etc
class ChatGPTClient {
  /// OpenAI ChatGPT Completions API Endpoint URL
  final url = Uri.https("api.openai.com", "/v1/chat/completions");

  /// OpenAI API Key which you can get from https://openai.com/api
  String apiKey;

  /// GPT Model (gpt-3.5-turbo, gpt-4, etc) default to gpt-3.5-turbo
  String model;

  /// System prompt, default to "You're a helpful assistant"
  String systemPrompt;

  /// Temperature, default to 0.5
  double temperature;

  List<Message> _historyList = List.empty(growable: true);

  /// Initializer, API key is required
  ChatGPTClient(
      {required this.apiKey,
      this.model = "gpt-3.5-turbo",
      this.systemPrompt = "You are a helpful assistant",
      this.temperature = 0.5});

  Map<String, String> _getHeaders() {
    return {
      "Content-Type": "application/json",
      "Authorization": "Bearer $apiKey"
    };
  }

  String _getBody(String text, bool stream) {
    final body = {
      "model": model,
      "temperature": temperature,
      "messages": _generateMessages(text).map((e) => e.toMap()).toList(),
      "stream": stream
    };
    return jsonEncode(body);
  }

  List<Message> _generateMessages(String prompt) {
    var messages = [_getSystemMessage()] +
        _historyList +
        [Message(content: prompt, role: "user")];
    final messagesContentCount = messages
        .map((e) => e.content.length)
        .reduce((value, element) => value + element);
    if (messagesContentCount > (4000 * 4)) {
      _historyList.removeAt(0);
      messages = _generateMessages(prompt);
    }
    return messages;
  }

  void _appendToHistoryList(String userText, String responseText) {
    _historyList.addAll([
      Message(content: userText, role: "user"),
      Message(content: responseText, role: "assistant")
    ]);
  }

  Message _getSystemMessage() {
    return Message(content: systemPrompt, role: "system");
  }

  /// Send message to ChatGPT to a prompt asynchronously
  Future<String> sendMessage(String text) async {
    final response = await http.Client()
        .post(url, headers: _getHeaders(), body: _getBody(text, false));

    dynamic decodedResponse;
    if (response.contentLength != null) {
      decodedResponse = jsonDecode(utf8.decode(response.bodyBytes)) as Map;
    }

    final statusCode = response.statusCode;
    if (!(statusCode >= 200 && statusCode < 300)) {
      if (decodedResponse != null) {
        final errorMessage = decodedResponse["error"]["message"] as String;
        throw Exception("($statusCode) $errorMessage");
      }
      throw Exception(
          "($statusCode) Bad response ${response.reasonPhrase ?? ""}");
    }

    final choices = decodedResponse["choices"] as List;
    final choice = choices[0] as Map;
    final content = choice["message"]["content"] as String;
    _appendToHistoryList(text, content);
    return content;
  }

  // Send Message to ChatGPT and receives the streamed response in chunk
  Stream<String> sendMessageStream(String text) async* {
    final request = http.Request("POST", url)..headers.addAll(_getHeaders());
    request.body = _getBody(text, true);

    final response = await StreamClient.instance.send(request);
    final statusCode = response.statusCode;
    final byteStream = response.stream;

    if (!(statusCode >= 200 && statusCode < 300)) {
      var error = "";
      await for (final byte in byteStream) {
        final decoded = utf8.decode(byte).trim();
        final map = jsonDecode(decoded) as Map;
        final errorMessage = map["error"]["message"] as String;
        error += errorMessage;
      }
      throw Exception(
          "($statusCode) ${error.isEmpty ? "Bad Response" : error}");
    }

    var responseText = "";
    await for (final byte in byteStream) {
      var decoded = utf8.decode(byte);
      final strings = decoded.split("data: ");
      for (final string in strings) {
        final trimmedString = string.trim();
        if (trimmedString.isNotEmpty && !trimmedString.endsWith("[DONE]")) {
          final map = jsonDecode(trimmedString) as Map;
          final choices = map["choices"] as List;
          final delta = choices[0]["delta"] as Map;
          if (delta["content"] != null) {
            final content = delta["content"] as String;
            responseText += content;
            yield content;
          }
        }
      }
    }

    if (responseText.isNotEmpty) {
      _appendToHistoryList(text, responseText);
    }
  }

  /// Clear history list array
  void clearHistoryList() {
    _historyList = List.empty(growable: true);
  }
}
